pkg/apiserver/watch.go

```diff
@@ -64,119 +59,218 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
+		s.scope.err(errors.NewInternalError(err), w, req)
+		return
+	}
+
+	// get a framed encoder
+	f, ok := s.encoder.(streaming.Framer)
+	if !ok {
+		// programmer error
+		err := fmt.Errorf("no streaming support is available for media type %q", s.mediaType)
+		utilruntime.HandleError(err)
+		s.scope.err(errors.NewBadRequest(err.Error()), w, req)
+		return
+	}
+	framer := f.NewFrameWriter(w)
+	if framer == nil {
+		// programmer error
+		err := fmt.Errorf("no stream framing support is available for media type %q", s.mediaType)
+		utilruntime.HandleError(err)
+		s.scope.err(errors.NewBadRequest(err.Error()), w, req)
 		return
 	}
+	e := streaming.NewEncoder(framer, s.encoder)
+
+	// ensure the connection times out
+	timeoutCh, cleanup := s.t.TimeoutCh()
+	defer cleanup()
+	defer s.watching.Stop()
+
+	// begin the stream
+	w.Header().Set("Content-Type", s.mediaType)
 	w.Header().Set("Transfer-Encoding", "chunked")
 	w.WriteHeader(http.StatusOK)
 	flusher.Flush()
-	// TODO: use arbitrary serialization on watch
-	encoder := watchjson.NewEncoder(w, self.encoder)
+
+	buf := &bytes.Buffer{}
 	for {
 		select {
 		case <-cn.CloseNotify():
 			return
 		case <-timeoutCh:
 			return
-		case event, ok := <-self.watching.ResultChan():
+		case event, ok := <-s.watching.ResultChan():
 			if !ok {
 				// End of results.
 				return
 			}
-			self.fixup(event.Object)
-			if err := encoder.Encode(&event); err != nil {
-				// Client disconnect.
+			obj := event.Object
+			s.fixup(obj)
+			if err := s.embeddedEncoder.EncodeToStream(obj, buf); err != nil {
+				// unexpected error
+				utilruntime.HandleError(fmt.Errorf("unable to encode watch object: %v", err))
+				return
+			}
+			event.Object = &runtime.Unknown{
+				Raw: buf.Bytes(),
+				// ContentType is not required here because we are defaulting to the serializer
+				// type
+			}
+			if err := e.Encode((*versioned.InternalEvent)(&event)); err != nil {
+				utilruntime.HandleError(fmt.Errorf("unable to encode watch object: %v", err))
+				// client disconnect.
 				return
 			}
 			flusher.Flush()
+
+			buf.Reset()
+		}
+	}
+}
+
+// HandleWS implements a websocket handler.
+func (s *WatchServer) HandleWS(ws *websocket.Conn) {
+	defer ws.Close()
+	done := make(chan struct{})
+	go wsstream.IgnoreReceives(ws, 0)
+	buf := &bytes.Buffer{}
+	streamBuf := &bytes.Buffer{}
+	for {
+		select {
+		case <-done:
+			s.watching.Stop()
+			return
+		case event, ok := <-s.watching.ResultChan():
+			if !ok {
+				// End of results.
+				return
+			}
+			obj := event.Object
+			s.fixup(obj)
+			if err := s.embeddedEncoder.EncodeToStream(obj, buf); err != nil {
+				// unexpected error
+				utilruntime.HandleError(fmt.Errorf("unable to encode watch object: %v", err))
+				return
+			}
+			event.Object = &runtime.Unknown{
+				Raw: buf.Bytes(),
+				// ContentType is not required here because we are defaulting to the serializer
+				// type
+			}
+			internalEvent := versioned.InternalEvent(event)
```

Comment:
The encoder is responsible for understanding how to convert an internal event to the proper format and version.  I can add a comment to that effect.